servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -598,21 +691,121 @@ public Completable onClose() {
                     ", #connections=" + connState.connections.length +
                     '}';
         }
+
+        private String describeState(Object state) {
+            if (ActiveState.class.equals(state.getClass())) {
+                return "active";
+            } else if (HealthCheck.class.equals(state.getClass())) {
+                return "unhealthy";
+            } else if (state == STATE_EXPIRED) {
+                return "expired";
+            } else if (state == STATE_CLOSED) {
+                return "closed";
+            } else {
+                return "unknown";
+            }
+        }
+
+        private static final class ActiveState {
+            private final int failedConnections;
+
+            ActiveState() {
+                this(0);
+            }
+
+            private ActiveState(int failedConnections) {
+                this.failedConnections = failedConnections;
+            }
+
+            ActiveState forNextFailedConnection() {
+                return new ActiveState(FlowControlUtils.addWithOverflowProtection(this.failedConnections, 1));
+            }
+        }
+
+        private static final class HealthCheck<ResolvedAddress, C extends LoadBalancedConnection>
+                extends DelayedCancellable {
+            private static final Exception RESCHEDULE_SIGNAL = ThrowableUtils.unknownStackTrace(
+                    new ConnectionRejectedException("Connection rejected during health check."),
+                    HealthCheck.class, "run()");
+            private final ConnectionFactory<ResolvedAddress, ? extends C> connectionFactory;
+            private final Host<ResolvedAddress, C> host;
+
+            private HealthCheck(final ConnectionFactory<ResolvedAddress, ? extends C> connectionFactory,
+                                final Host<ResolvedAddress, C> host) {
+                this.connectionFactory = connectionFactory;
+                this.host = host;
+            }
+
+            public void schedule() {
+                assert host.healthCheckConfig != null;
+                delayedCancellable(
+                        retryWithConstantBackoffFullJitter(cause -> true,
+                                host.healthCheckConfig.healthCheckInterval,
+                                host.healthCheckConfig.executor)
+                                .apply(0, RESCHEDULE_SIGNAL)
+                                .concat(reconnect()
+                                        .retryWhen(retryWithConstantBackoffFullJitter(
+                                                cause -> cause == RESCHEDULE_SIGNAL,
+                                                host.healthCheckConfig.healthCheckInterval,
+                                                host.healthCheckConfig.executor)))
+                                .subscribe());
+            }
+
+            public Completable reconnect() {
+                return defer(() -> connectionFactory.newConnection(host.address, null))
```

Comment:
Doesn't look like `defer` is required here

Reply:
As a matter of fact it is. It might feel it isn't and in most cases there won't be any issue. But the fact that the call `connectionFactory.newConnection(host.address, null)` can return a different `Single` every time it's called makes the re-subscribing mechanism prone to errors if that call is not re-exercised upon every retry. A practical example can be observed when you run the RRLB tests after removing the `defer()` call. But aside from the quirks of the test, I just believe it's semantically incorrect to assume the factory returns a proper re-subscribable `Single` that would behave the same way when a new one was obtained via calling `newConnection`.

Reply:
In Reactive Streams each re-subscribe should trigger new evaluation of the `Single` and produce the meaningful result. The result might be different (like a connection with new id), but it's still a result that should be treated as any previous result. In this case, your test is written incorrectly, without properly following RS spec:
```java
        Function<String, Single<TestLoadBalancedConnection>> factory =
                new Function<String, Single<TestLoadBalancedConnection>>() {

            @Override
            public Single<TestLoadBalancedConnection> apply(final String s) {
                if (s.equals(failingHost)) {
                    requests.incrementAndGet();
                    if (momentInTime.get() >= connections.size()) {
                        return properConnection;
                    }
                    return connections.get(momentInTime.get());
                }
                return properConnection;
            }
        };
```
Your `apply` method returns a `Single`, but it creates a state outside of the `Single`, which is not executed on each re-subscribe. Look at any other filter. For example, `RequestTargetEncoderHttpRequesterFilter`. It uses `defer` to recompute `requestTarget` on each re-subscribe. Otherwise, retry or repeat operator won't work as expected. You should do the same here and wrap the internals of your `apply` function with defer instead of using `defer` inside RRLB.

Reply:
That's a great explanation, thanks for taking the time to elaborate. That makes sense, I did as you recommended.